File: backend/app/schemas/weather.py
from typing import Optional, List, Dict, Any, Union
from pydantic import BaseModel, Field
from backend.app.schemas.alerts import IMDAlert

class LocationMeta(BaseModel):
    latitude: float
    longitude: float
    city: Optional[str] = None
    district: Optional[str] = None
    state: Optional[str] = None
    country: Optional[str] = "India"
    imd_station: Optional[str] = None
    imd_subdivision: Optional[str] = None

class CurrentWeather(BaseModel):
    temperature: Optional[float] = None
    feels_like: Optional[float] = None
    humidity: Optional[float] = None
    precipitation: Optional[float] = None
    rain_probability: Optional[float] = None
    wind_speed: Optional[float] = None
    wind_direction: Optional[Union[str, float]] = None
    pressure: Optional[float] = None
    visibility: Optional[float] = None
    uv_index: Optional[float] = None
    condition: str = "Clear Sky"
    icon: Optional[str] = "sun"
    weather_code: Optional[int] = 0
    observed_at: str
    source: str = "Open-Meteo"

class HourlyForecastItem(BaseModel):
    time: str
    iso_time: Optional[str] = None
    temp: Optional[float] = None
    condition: str
    icon: str
    rainProb: Optional[float] = 0
    highlight: Optional[bool] = False
    source: str = "Open-Meteo"
    weather_code: Optional[int] = None
    precipitation: Optional[float] = None
    wind_speed: Optional[float] = None

class DailyForecastItem(BaseModel):
    day: str
    date: str
    high: Optional[float] = None
    low: Optional[float] = None
    condition: str
    icon: str
    rainProbability: Optional[float] = 0
    humidity: Optional[float] = None
    source: str = "Open-Meteo"

class WeatherSource(BaseModel):
    provider: str = "Open-Meteo"
    retrieved_at: str
    is_cached: bool = False

class IMDOfficialInformation(BaseModel):
    source: str = "India Meteorological Department"
    status: str = "available"  # "available" | "unavailable"
    station_name: Optional[str] = None
    station_code: Optional[str] = None
    district: Optional[str] = None
    subdivision: Optional[str] = None
    observed_at: Optional[str] = None
    temperature: Optional[float] = None
    humidity: Optional[float] = None
    rainfall_past_24h: Optional[float] = None
    forecast_summary: Optional[str] = None
    attribution: str = "Official Data Source: India Meteorological Department, Ministry of Earth Sciences, Govt. of India"
    message: Optional[str] = None

class WeatherData(BaseModel):
    location: LocationMeta
    current: CurrentWeather
    hourly: List[HourlyForecastItem] = []
    daily: List[DailyForecastItem] = []
    source: WeatherSource
    official_information: Optional[IMDOfficialInformation] = None
    alerts: List[IMDAlert] = []

class WeatherApiResponse(BaseModel):
    success: bool = True
    data: Optional[WeatherData] = None
    message: Optional[str] = None

File: frontend/src/types/weather.ts
import { UserLocation } from './location';
import { IMDAlert } from './alert';

export type { UserLocation, IMDAlert };

export type WeatherCondition =
  | 'Sunny'
  | 'Partly Cloudy'
  | 'Cloudy'
  | 'Overcast'
  | 'Light Rain'
  | 'Moderate Rain'
  | 'Heavy Rain'
  | 'Thunderstorm'
  | 'Showers'
  | string;

export interface WeatherLocationMeta {
  latitude: number;
  longitude: number;
  city: string;
  district: string;
  state: string;
  country: string;
  source?: string;
  imd_station?: string;
  imd_subdivision?: string;
}

export interface WeatherCurrent {
  temperature: number | null;
  feels_like: number | null;
  humidity: number | null;
  rain_probability: number | null;
  precipitation: number | null;
  wind_speed: number | null;
  wind_direction: string | null;
  visibility: number | null;
  pressure: number | null;
  uv_index: number | null;
  condition: string;
  icon: string;
  weather_code: number | null;
  observed_at: string | null;
  source?: string;
}

export interface WeatherHourlyItem {
  time: string;
  iso_time?: string;
  temp: number;
  condition: string;
  icon: string;
  rainProb: number;
  highlight?: boolean;
  source?: string;
}

export interface WeatherDailyItem {
  day: string;
  date: string;
  high: number;
  low: number;
  condition: string;
  icon: string;
  rainProbability: number;
  rainProb?: number;
  humidity: number;
  source?: string;
}

export interface WeatherSourceMeta {
  provider: string;
  retrieved_at: string | null;
  is_cached?: boolean;
}

export interface IMDOfficialInformation {
  source: string;
  status: 'available' | 'unavailable' | string;
  station_name?: string | null;
  station_code?: string | null;
  district?: string | null;
  subdivision?: string | null;
  observed_at?: string | null;
  temperature?: number | null;
  humidity?: number | null;
  rainfall_past_24h?: number | null;
  forecast_summary?: string | null;
  attribution?: string;
  message?: string | null;
}

export interface WeatherGPTResponse {
  location: WeatherLocationMeta;
  current: WeatherCurrent;
  hourly: WeatherHourlyItem[];
  daily: WeatherDailyItem[];
  source: WeatherSourceMeta;
  official_information?: IMDOfficialInformation;
  alerts?: IMDAlert[];
}

// Legacy interfaces retained for backward compatibility with secondary components
export interface CurrentWeather {
  temperature: number;
  feelsLike: number;
  condition: string;
  humidity: number;
  windSpeed: number;
  windDirection: string;
  visibility: number;
  pressure: number;
  uvIndex: number;
  rainProbability: number;
  updatedAt: string;
  highTemp?: number;
  lowTemp?: number;
  airQualityIndex?: number;
  airQualityLabel?: string;
}

export interface HourlyForecast {
  time: string;
  temperature: number;
  condition: string;
  rainProbability: number;
  isHighRisk?: boolean;
}

export interface DailyForecast {
  day: string;
  date: string;
  high: number;
  low: number;
  condition: string;
  rainProbability: number;
  humidity: number;
}

export interface HyperlocalRisk {
  areaName: string;
  distanceKm: number;
  rainRisk: 'LOW' | 'MEDIUM' | 'HIGH' | 'EXTREME';
  waterloggingRisk: 'LOW' | 'MEDIUM' | 'HIGH' | 'EXTREME';
  lightning: 'LOW' | 'MEDIUM' | 'HIGH' | 'EXTREME';
  wind: 'LOW' | 'MEDIUM' | 'HIGH' | 'EXTREME';
  temperature: number;
}
